### PR TITLE
Implement VonMises3D distribution

### PR DESCRIPTION
**Pull request by fritzo**
This implements the `.log_prob()` part of a `VonMises3D` distribution, similar to the 2D `VonMises` distribution.

This also fixes an attribute name bug (my fault) introduced in #1187.

## Tested

- [x] simple normalization test
- [x] tested on internal applications
- [ ] any other tests?

### CODE REVIEW

**Comment by alicanb** [comment]
Do you think it's worthwhile to merge `VonMises` and `VonMises3D`?

**Comment by fritzo** [comment]
Do you mean "put them in the same file"? I think they're different enough to warrant separate classes, since we use polar coords for `VonMises` and cartesian coords for `VonMises3D`.

File: pyro/distributions/von_mises_3d.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+class VonMises3D(TorchDistribution):
+    """
+    Spherical von Mises distribution.
+    https://en.wikipedia.org/wiki/Von_Mises%E2%80%93Fisher_distribution
+
+    Currently only :meth:`log_prob` is implemented.
+
+    :param torch.Tensor concentration: A combined location-and-concentration
+        vector. The direction of this vector is the location, and its
+        magnitude is the contration.
```
Comment (alicanb) [inline pyro/distributions/von_mises_3d.py]: contration -> concentration

File: pyro/distributions/von_mises.py [hunk 2]
```diff
@@ -68,7 +68,7 @@ def expand(self, batch_shape):
         try:
             return super(VonMises, self).expand(batch_shape)
         except NotImplementedError:
-            validate_args = self.__dict__.get('validate_args')
+            validate_args = self.__dict__.get('_validate_args')
```
Comment (neerajprad) [inline pyro/distributions/von_mises.py]: Thanks for fixing this, and sorry I should have caught this earlier!

File: tests/distributions/test_von_mises.py [hunk 3]
```diff
@@ -14,3 +14,19 @@ def test_log_prob_normalized(concentration):
...
+@pytest.mark.parametrize('scale', [0.1, 0.5, 0.9, 1.0, 1.1, 2.0, 10.0])
+def test_von_mises_3d(scale):
+    concentration = torch.randn(3)
+    concentration = concentration * (scale / concentration.norm(2))
+
+    num_samples = 100000
+    samples = torch.randn(num_samples, 3)
+    samples = samples / samples.norm(2, dim=-1, keepdim=True)
+
+    d = VonMises3D(concentration, validate_args=True)
```
Comment (neerajprad) [inline tests/distributions/test_von_mises.py]: If we do not explicitly turn this flag to `True`, is the distributions arg checking disabled? We should have it enabled by default, unless there has been some regression lately.

File: pyro/distributions/von_mises_3d.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+import torch
+from torch.distributions import constraints
+
+from pyro.distributions import TorchDistribution
+
+
+class VonMises3D(TorchDistribution):
+    """
+    Spherical von Mises distribution.
+    https://en.wikipedia.org/wiki/Von_Mises%E2%80%93Fisher_distribution
```
Comment (jpchen) [inline pyro/distributions/von_mises_3d.py]: nit: this would be better as a hyperlink. also we should have this for all distributions or none imo

File: tests/distributions/test_von_mises.py (hunk 3 above)
Comment (fritzo) [reply tests/distributions/test_von_mises.py]: Good point, I'll update this.

**Comment by fritzo** [comment]
> we could probably bake some logic over which Implementation to use

No, `VonMises` is in polar coords whereas `VonMises3D` is in Cartesian submanifold coords. If we ever generalize, it would be from `VonMises3D` to `VonMisesFisher` or something that would use Cartesian coords for all dimensions.

I'll add comments explaining this in both distributions, since it has already caused confusion twice.

File: pyro/distributions/von_mises_3d.py (hunk 4 above)
Comment (fritzo) [reply pyro/distributions/von_mises_3d.py]: Sure, I'll add explanation and remove the link.

**Comment by jpchen** [comment]
> VonMises is in polar coords whereas VonMises3D is in Cartesian submanifold coords.

ah i missed the discussion above, yeah i guess in that case it's better to fail than silently do the wrong thing, though i usually assume the user is cognizant of the nuances of args s/he passes to distributions
